---
title: "STAR"
output: md_document
---

```{r setup, include=FALSE}
Sys.setenv(PATH=paste0(Sys.getenv("PATH"), ":", getwd()))
knitr::opts_chunk$set(echo = TRUE)
```

STAR usage can be accessed using `--help`. (Since the help page is quite long, only the first 10 lines are shown below.)

```{bash engine.opts='-l'}
STAR --help | head
```

Generate 100 unique reads with read lengths of 100 bp and repeat each unique read incrementally, i.e. first read is unique, second read is repeated twice, third read is repeated thrice, and so on. Next use these reads to generate a reference sequence where each read is separated by a spacer sequence that is two times the length of a read, which is 200 bp in this case.

The reference sequence is first indexed and then the reads are mapped to the reference using default settings. We will set `--genomeSAindexNbases` based on the recommendation:

>For small genomes, the parameter --genomeSAindexNbases must be scaled down to min(14, log2(GenomeLength)/2 - 1).

```{bash engine.opts='-l'}
l=100
r=100
../script/create_ref_se.pl -l ${l} -r ${r}

genome=$(cat l${l}_r${r}_ref.fa | grep -v "^>" | wc -m)
param=$(printf "%.f" $(bc -l<<<"(l(${genome}) / l(2)) / 2 - 1"))

STAR \
   --runMode genomeGenerate \
   --genomeDir star_idx \
   --genomeFastaFiles l${l}_r${r}_ref.fa \
   --genomeSAindexNbases ${param} \
   --runThreadN 8 > /dev/null

STAR --runMode alignReads \
   --genomeDir star_idx \
   --readFilesIn l${l}_r${r}_reads.fa \
   --outFileNamePrefix default. > /dev/null
```

STAR will map to up to 10 loci by default.

> outFilterMultimapNmax           10
>     int: maximum number of loci the read is allowed to map to. Alignments (all of them) will be output only if the read maps to no more loci than this value.
>          Otherwise no alignments will be output, and the read will be counted as "mapped to too many loci" in the Log.final.out .

Any read mapping to more than 10 loci will become unmapped and will be excluded from the output SAM by default as per the `outSAMunmapped` parameter. To include unmapped reads use `outSAMunmapped Within` (not used in this README).

> outSAMunmapped                  None
>     string(s): output of unmapped reads in the SAM format
>                                 1st word:
>                                 None   ... no output
>                                 Within ... output unmapped reads within the main SAM file (i.e. Aligned.out.sam)
>                                 2nd word:
>                                 KeepPairs ... record unmapped mate for each alignment, and, in case of unsorted output, keep it adjacent to its mapped mate. Only affects multi-mapping reads.


```{bash engine.opts='-l'}
cat default.Aligned.out.sam | grep -v "^@" | cut -f1 | sort | uniq -c | sort -k2n
```

Even with `outFilterMultimapNmax` set to 100, the maximum number of loci is 50.

```{bash engine.opts='-l'}
l=100
r=100

STAR --runMode alignReads \
   --genomeDir star_idx \
   --readFilesIn l${l}_r${r}_reads.fa \
   --outFilterMultimapNmax 100 \
   --outFileNamePrefix m100. > /dev/null

cat m100.Aligned.out.sam | grep -v "^@" | cut -f1 | sort | uniq -c | sort -k2rn | head
```

Setting `winAnchorMultimapNmax` to 100 does not increase the number of multimappers.

> winAnchorMultimapNmax           50
>     int>0: max number of loci anchors are allowed to map to

```{bash engine.opts='-l'}
l=100
r=100

STAR --runMode alignReads \
   --genomeDir star_idx \
   --readFilesIn l${l}_r${r}_reads.fa \
   --outFilterMultimapNmax 100 \
   --winAnchorMultimapNmax 100 \
   --outFileNamePrefix ma100. > /dev/null

cat ma100.Aligned.out.sam | grep -v "^@" | cut -f1 | sort | uniq -c | sort -k2rn | head
```

Increasing `limitOutSAMoneReadBytes` also does not increase the number of multimappers.

> limitOutSAMoneReadBytes              100000
>     int>0: max size of the SAM record (bytes) for one read. Recommended value: >(2*(LengthMate1+LengthMate2+100)*outFilterMultimapNmax

```{bash engine.opts='-l'}
l=100
r=100

STAR --runMode alignReads \
   --genomeDir star_idx \
   --readFilesIn l${l}_r${r}_reads.fa \
   --outFilterMultimapNmax 100 \
   --winAnchorMultimapNmax 100 \
   --limitOutSAMoneReadBytes 1000000 \
   --outFileNamePrefix ml100. > /dev/null

cat ml100.Aligned.out.sam | grep -v "^@" | cut -f1 | sort | uniq -c | sort -k2rn | head
```

Increasing `--winAnchorMultimapNmax` and `seedPerWindowNmax` to 100 increases the number of loci a read can map to.

> seedPerWindowNmax     50
>     int>0: max number of seeds per window

```{bash engine.opts='-l'}
l=100
r=100

STAR --runMode alignReads \
   --genomeDir star_idx \
   --readFilesIn l${l}_r${r}_reads.fa \
   --outFilterMultimapNmax 100 \
   --winAnchorMultimapNmax 100 \
   --seedPerWindowNmax 100 \
   --outFileNamePrefix ms100. > /dev/null

cat ms100.Aligned.out.sam | grep -v "^@" | cut -f1 | sort | uniq -c | sort -k2rn | head
```

Regarding the order of multimappers, there is the `outMultimapperOrder` parameter:

> outMultimapperOrder             Old_2.4
>     string: order of multimapping alignments in the output files
>                                 Old_2.4             ... quasi-random order used before 2.5.0
>                                 Random              ... random order of alignments for each multi-mapper. Read mates (pairs) are always adjacent, all alignment for each read stay together. This option will become default in the future releases.

However it seems like the same loci is chosen as the primary alignment.

```{bash engine.opts='-l'}
l=100
r=100

STAR --runMode alignReads \
   --genomeDir star_idx \
   --readFilesIn l${l}_r${r}_reads.fa \
   --outFileNamePrefix default2. > /dev/null

STAR --runMode alignReads \
   --genomeDir star_idx \
   --readFilesIn l${l}_r${r}_reads.fa \
   --outFileNamePrefix default3. > /dev/null

STAR --runMode alignReads \
   --genomeDir star_idx \
   --readFilesIn l${l}_r${r}_reads.fa \
   --outFileNamePrefix default4. > /dev/null

cat default*.sam | grep "^5"
```

Map repetitive reads to chrX.

```{bash engine.opts='-l'}
gunzip -c ../data/chrX.fa.gz > chrX.fa

STAR \
   --runMode genomeGenerate \
   --genomeDir chrx \
   --genomeFastaFiles chrX.fa \
   --genomeSAindexNbases 12 \
   --runThreadN 8 > /dev/null

STAR --runMode alignReads \
   --genomeDir chrx \
   --outFilterMultimapNmax 1000 \
   --winAnchorMultimapNmax 1000 \
   --seedPerWindowNmax 1000 \
   --readFilesCommand "gunzip -c" \
   --readFilesIn ../data/chrx_kmer.fa.gz \
   --outFileNamePrefix chrx. \
   --runThreadN 8 > /dev/null
```

Reads repeating up to 10 times are mapped as expected.

```{bash engine.opts='-l'}
cat chrx.Aligned.out.sam | grep -v "^@" | cut -f1 | sort | uniq -c | sort -k2n | head
```

However reads repeating up to 1,000 times are not mapped to all their loci. In addition, note that reads mapping over 1,000 times are discarded (therefore STAR does map up to 1,000 times but it just does not report all the loci with the provided parameters).

```{bash engine.opts='-l'}
cat chrx.Aligned.out.sam | grep -v "^@" | cut -f1 | sort | uniq -c | sort -k2rn | head
```

Clean up.

```{bash engine.opts='-l'}
rm -rf *.sam *.fa star_idx chrx *.out *.out.tab
```

STAR version used for this README.

```{bash engine.opts='-l'}
STAR --version
```

